editor/plugins/theme_editor_plugin.cpp

```diff
@@ -31,12 +31,665 @@
+			edit_theme_item_dialog->set_title(TTR("Add Stylebox Item"));
+			break;
+		case Theme::DATA_TYPE_MAX:
+			break; // Can't happen, but silences warning.
+	}
+
+	edit_theme_item_old_vb->hide();
+	theme_item_name->clear();
+	edit_theme_item_dialog->popup_centered(Size2(280, 140) * EDSCALE);
+	theme_item_name->grab_focus();
+}
+
+void ThemeItemEditorDialog::_open_rename_theme_item_dialog(Theme::DataType p_data_type, String p_item_name) {
+	ERR_FAIL_COND_MSG(p_data_type < 0 || p_data_type >= Theme::DATA_TYPE_MAX, "Theme item data type is out of bounds.");
+
+	item_popup_mode = RENAME_THEME_ITEM;
+	edit_item_data_type = p_data_type;
+	edit_item_old_name = p_item_name;
+
+	switch (edit_item_data_type) {
+		case Theme::DATA_TYPE_COLOR:
+			edit_theme_item_dialog->set_title(TTR("Rename Color Item"));
+			break;
+		case Theme::DATA_TYPE_CONSTANT:
+			edit_theme_item_dialog->set_title(TTR("Rename Constant Item"));
+			break;
+		case Theme::DATA_TYPE_FONT:
+			edit_theme_item_dialog->set_title(TTR("Rename Font Item"));
+			break;
+		case Theme::DATA_TYPE_FONT_SIZE:
+			edit_theme_item_dialog->set_title(TTR("Rename Font Size Item"));
+			break;
+		case Theme::DATA_TYPE_ICON:
+			edit_theme_item_dialog->set_title(TTR("Rename Icon Item"));
+			break;
+		case Theme::DATA_TYPE_STYLEBOX:
+			edit_theme_item_dialog->set_title(TTR("Rename Stylebox Item"));
+			break;
+		case Theme::DATA_TYPE_MAX:
+			break; // Can't happen, but silences warning.
+	}
+
+	edit_theme_item_old_vb->show();
+	theme_item_old_name->set_text(p_item_name);
+	theme_item_name->set_text(p_item_name);
+	edit_theme_item_dialog->popup_centered(Size2(280, 140) * EDSCALE);
+	theme_item_name->grab_focus();
+}
+
+void ThemeItemEditorDialog::_confirm_edit_theme_item() {
+	if (item_popup_mode == CREATE_THEME_ITEM) {
+		_add_theme_item(edit_item_data_type, theme_item_name->get_text(), edited_item_type);
+	} else if (item_popup_mode == RENAME_THEME_ITEM) {
+		edited_theme->rename_theme_item(edit_item_data_type, edit_item_old_name, theme_item_name->get_text(), edited_item_type);
+	}
+
+	item_popup_mode = ITEM_POPUP_MODE_MAX;
+	edit_item_data_type = Theme::DATA_TYPE_MAX;
+	edit_item_old_name = "";
+
+	_update_edit_item_tree(edited_item_type);
+}
+
+void ThemeItemEditorDialog::_edit_theme_item_gui_input(const Ref<InputEvent> &p_event) {
+	Ref<InputEventKey> k = p_event;
+
+	if (k.is_valid()) {
+		if (!k->is_pressed()) {
+			return;
+		}
+
+		switch (k->get_keycode()) {
+			case KEY_KP_ENTER:
+			case KEY_ENTER: {
+				_confirm_edit_theme_item();
+				edit_theme_item_dialog->hide();
+				edit_theme_item_dialog->set_input_as_handled();
+			} break;
+			case KEY_ESCAPE: {
+				edit_theme_item_dialog->hide();
+				edit_theme_item_dialog->set_input_as_handled();
+			} break;
+		}
+	}
+}
+
+void ThemeItemEditorDialog::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_ENTER_TREE: {
+			connect("about_to_popup", callable_mp(this, &ThemeItemEditorDialog::_dialog_about_to_show));
```

Comment:
Why is this not being connected in the constructor?

Reply:
Because putting it in the constructor results in an error

```cpp
ThemeItemEditorDialog::ThemeItemEditorDialog() {
	set_title(TTR("Edit Theme Items"));
	connect("about_to_popup", callable_mp(this, &ThemeItemEditorDialog::_dialog_about_to_show));
```

```
ERROR: In Object of type 'ThemeItemEditorDialog': Attempt to connect nonexistent signal 'about_to_popup' to callable 'ThemeItemEditorDialog::_dialog_about_to_show'.
   at: (core\object\object.cpp:1274)
```

I assumed that's something to do with the order in which inherited properties are applied. Unless I did something else wrong?

Reply:
Ok this is actually correct. I got confused, because calling `connect` on other objects is more common and it's always done in the constructor.